### PR TITLE
Jenkins: Make mysql and postgres tests run in parallel

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

Run the mysql and postgres stages inside parallel blocks. Essentially the same thing done in Jenkinsfile.branch for build EE and build TE.

#### Ticket link
https://mattermost.atlassian.net/browse/MM-23892

### CODE REVIEW

File: build/Jenkinsfile.pr [hunk 1]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
-							export MM_SQLSETTINGS_DATASOURCE=\"postgres://mmuser:mostest@postgres:5432/latest?sslmode=disable&connect_timeout=10\"
-							make ARGS="config get SqlSettings.DataSource" run-cli
-
-							echo "Setting up fresh db"
-							make ARGS="version" run-cli
-						"""
-					}
-				}
 
+					steps {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: I think we can extract this to the stage "Test" as a common part that is not run in parallel

File: build/Jenkinsfile.pr [hunk 2]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
+							}
+						}
 
-				dir('src/github.com/mattermost/mattermost-server') {
-					ansiColor('xterm') {
-						sh """
-						echo "Generating dump"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T postgres pg_dump --schema-only -d migrated -U mmuser > migrated.sql
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T postgres pg_dump --schema-only -d latest -U mmuser > latest.sql
+						dir('src/github.com/mattermost/mattermost-server') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: This looks like only postgres related

File: build/Jenkinsfile.pr [hunk 3]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
+								/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T postgres sh -c 'exec echo "CREATE DATABASE migrated; CREATE DATABASE latest;" | exec psql -U mmuser mattermost_test'
+								echo "Importing postgres dump from version 5.0"
+								/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T postgres psql -U mmuser -d migrated < \$(pwd)/scripts/mattermost-postgresql-5.0.sql
+								"""
+							}
+						}
 
-						echo "Removing databases created for db comparison"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T postgres sh -c \'exec echo \"DROP DATABASE migrated; DROP DATABASE latest;\" | exec psql -U mmuser mattermost_test\'
+						withDockerContainer(args: "-u root --privileged --net ${COMPOSE_PROJECT_NAME}_mm-test -v ${WORKSPACE}/src:/go/src/", image: 'mattermost/mattermost-build-server:oct-18-2019') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: again, looks like postgres related only.

File: build/Jenkinsfile.pr [hunk 4]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
+						}
 
-						echo "Generating diff"
-						set +e
-						diff migrated.sql latest.sql > diff.txt
-						set -e
-						export diffErrorCode=\$?
 
-						if [ \$diffErrorCode -eq 0 ]; then echo \"Both schemas are same\";else cat diff.txt; fi
+						dir('src/github.com/mattermost/mattermost-server') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: Again, postgres only

File: build/Jenkinsfile.pr [hunk 5]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
-							make ARGS="version" run-cli
-						"""
-					}
-				}
+								exit \$diffErrorCode
+								"""
+							}
+						}
 
+						dir('src/github.com/mattermost/mattermost-server') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: looks like related only to mysql.

File: build/Jenkinsfile.pr [hunk 6]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
+							}
+						}
 
-				dir('src/github.com/mattermost/mattermost-server') {
-					ansiColor('xterm') {
-						sh """
-						echo "Ignoring known MySQL mismatch: ChannelMembers.SchemeGuest"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml exec -T mysql mysql -D migrated -uroot -pmostest -e "ALTER TABLE ChannelMembers DROP COLUMN SchemeGuest;"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml exec -T mysql mysql -D latest -uroot -pmostest -e "ALTER TABLE ChannelMembers DROP COLUMN SchemeGuest;"
+						withDockerContainer(args: "-u root --privileged --net ${COMPOSE_PROJECT_NAME}_mm-test -v ${WORKSPACE}/src:/go/src/", image: 'mattermost/mattermost-build-server:oct-18-2019') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: looks like related only to mysql

File: build/Jenkinsfile.pr [hunk 7]
```diff
@@ -237,336 +237,340 @@ pipeline {
...
+							}
+						}
 
-						echo "Generating dump"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T mysql mysqldump --skip-opt --no-data --compact -u root -pmostest migrated > migrated.sql
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T mysql mysqldump --skip-opt --no-data --compact -u root -pmostest latest > latest.sql
 
-						echo "Removing databases created for db comparison"
-						/usr/local/bin/docker-compose --no-ansi -f build/docker-compose.yml  exec -T mysql mysql -uroot -pmostest -e \"DROP DATABASE migrated; DROP DATABASE latest\"
+						dir('src/github.com/mattermost/mattermost-server') {
```
Comment (jespino) [inline build/Jenkinsfile.pr]: looks like related only to mysql

**Comment by agnivade** [comment]
> tests are failing and some settings are probably wrongly duplicated (from the previous existing version).

Yeah, that's what I was surprised to see. How is it even working now ?

**Comment by agnivade** [comment]
This is ready for review now ! I have separated the docker-compose builds by giving each of them a unique `COMPOSE_PROJECT_NAME`.

My only remaining concern is with the `post, always` stage. Does `docker-compose logs --tail="all"` gather logs from containers under all projects ? If yes, then I believe this is good to go.

**Comment by agnivade** [comment]
I am finding a lot of flakes by running Jenkins like this. Not sure why :/ There doesn't seem to be anything wrong with the setup. Just regular tests are failing.

Sent 1 here: https://github.com/mattermost/enterprise/pull/605

**Comment by agnivade** [comment]
Looks like related to openldap setup.

**Comment by agnivade** [comment]
@metanerd @jespino - I think this is ready for review now. After the Jenkins issue got fixed, it now runs fine. I tested it twice, and one it just failed with a flaky test. There is no issue with the parallel test setup.

I just made one change from the original. viz. now I am setting `allowEmptyResults` to false in the junit command rather than true. This is what was recommended [here](https://jenkins.io/doc/pipeline/steps/junit/).

> If checked, the default behavior of failing a build on missing test result files or empty test results is changed to not affect the status of the build. Please note that this setting make it harder to spot misconfigured jobs or build failures where the test tool does not exit with an error code when not producing test report files.

**Comment by agnivade** [comment]
@lindy65 - Just FYI that this was a Jenkins improvement and is not code related, so no QA required. I am going ahead and merging it.

**Comment by lindy65** [comment]
Thanks @agnivade - just added the QA Review Done label for completeness ;)

**Pull request merged**